File: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/dominio/inss/multa/TaxaMultaPrevidenciaria.java
/*
 * Decompiled with CFR 0.152.
 * 
 * Could not load the following classes:
 *  javax.persistence.Column
 *  javax.persistence.Entity
 *  javax.persistence.GeneratedValue
 *  javax.persistence.GenerationType
 *  javax.persistence.Id
 *  javax.persistence.SequenceGenerator
 *  javax.persistence.Table
 *  javax.persistence.Temporal
 *  javax.persistence.TemporalType
 *  org.hibernate.annotations.Parameter
 *  org.hibernate.annotations.Type
 */
package br.jus.trt8.pjecalc.negocio.dominio.inss.multa;

import br.jus.trt8.pjecalc.base.comum.Competencia;
import br.jus.trt8.pjecalc.base.comum.EntidadeBase;
import br.jus.trt8.pjecalc.base.comum.HelperDate;
import br.jus.trt8.pjecalc.base.comum.Utils;
import br.jus.trt8.pjecalc.negocio.constantes.TipoDaMultaDoINSSEnum;
import br.jus.trt8.pjecalc.negocio.constantes.TipoDeAliquotaDoEmpregadorEnum;
import br.jus.trt8.pjecalc.negocio.constantes.TipoDeAliquotaDoSeguradoEnum;
import br.jus.trt8.pjecalc.negocio.constantes.TipoPagamentoDaMultaDoINSSEnum;
import br.jus.trt8.pjecalc.negocio.constantes.TipoPeriodicidadeMultaPrevidenciariaEnum;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.Calculo;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.atualizacao.ParametrosDeAtualizacao;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.inss.sobresalarios.InssSobreSalarios;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.inss.sobresalarios.InssSobreSalariosDevidos;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.inss.sobresalarios.InssSobreSalariosPagos;
import br.jus.trt8.pjecalc.negocio.dominio.calculoexterno.ParcelasAtualizaveisOutrosDebitosReclamado;
import br.jus.trt8.pjecalc.negocio.dominio.inss.atividadeeconomica.AtividadeEconomica;
import br.jus.trt8.pjecalc.negocio.dominio.inss.multa.RepositorioDeTaxaMultaPrevidenciaria;
import br.jus.trt8.pjecalc.negocio.dominio.inss.multa.TaxaMultaPrevidenciariaOptimizerListSearch;
import java.math.BigDecimal;
import java.util.Collection;
import java.util.Date;
import java.util.List;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.SequenceGenerator;
import javax.persistence.Table;
import javax.persistence.Temporal;
import javax.persistence.TemporalType;
import org.hibernate.annotations.Parameter;
import org.hibernate.annotations.Type;

@Entity
@Table(name="TBTAXAMULTAINSS")
@SequenceGenerator(name="SQTAXAMULTAINSS", sequenceName="SQTAXAMULTAINSS", allocationSize=1)
public class TaxaMultaPrevidenciaria
extends EntidadeBase {
    private static final long serialVersionUID = -5781080650751597967L;
    private static final HelperDate TRINTA_SETEMBRO_2008 = HelperDate.getInstance(2008, 8, 30);
    private static final String SERVICOS_DOMESTICOS = "SERVI\u00c7OS DOM\u00c9STICOS";
    private static final BigDecimal TAXA_MULTA_CALCULO_EXTERNO = new BigDecimal("0.33");
    private static final BigDecimal TETO_MULTA_CALCULO_EXTERNO = new BigDecimal("20");
    @Id
    @GeneratedValue(strategy=GenerationType.SEQUENCE, generator="SQTAXAMULTAINSS")
    @Column(name="IIDTAXAMULTA", nullable=false)
    private final Long id = null;
    @Column(name="DDTINICIO")
    @Temporal(value=TemporalType.DATE)
    private Date dataInicial;
    @Column(name="DDTFIM")
    @Temporal(value=TemporalType.DATE)
    private Date dataFinal;
    @Column(name="STPMULTA", columnDefinition="VARCHAR2(1)")
    @Type(type="br.jus.trt8.pjecalc.base.comum.EnumUserType", parameters={@Parameter(name="enum", value="TipoDaMultaDoINSSEnum")})
    private TipoDaMultaDoINSSEnum tipoMulta = TipoDaMultaDoINSSEnum.URBANA;
    @Column(name="STPPERIODICIDADE", columnDefinition="VARCHAR2(1)")
    @Type(type="br.jus.trt8.pjecalc.base.comum.EnumUserType", parameters={@Parameter(name="enum", value="TipoPeriodicidadeMultaPrevidenciariaEnum")})
    private TipoPeriodicidadeMultaPrevidenciariaEnum periodicidadeMulta = TipoPeriodicidadeMultaPrevidenciariaEnum.TAXA_UNICA;
    @Column(name="RVLTAXA", precision=38, scale=25)
    private BigDecimal taxa;
    @Column(name="RVLTETOTAXADIARIA", precision=38, scale=25)
    private BigDecimal tetoTaxaDiaria;
    @Column(name="SFLPERMITEREDUCAO", nullable=false, columnDefinition="VARCHAR2(1)")
    @Type(type="br.jus.trt8.pjecalc.base.comum.BooleanUserType")
    private Boolean permiteReducao = Boolean.FALSE;

    public TaxaMultaPrevidenciaria() {
        super(RepositorioDeTaxaMultaPrevidenciaria.class);
    }

    @Override
    public Object obterChavePrimaria() {
        return this.getId();
    }

    public Long getId() {
        return this.id;
    }

    public Date getDataInicial() {
        return this.dataInicial;
    }

    public Date getDataFinal() {
        return this.dataFinal;
    }

    public TipoDaMultaDoINSSEnum getTipoMulta() {
        return this.tipoMulta;
    }

    public TipoPeriodicidadeMultaPrevidenciariaEnum getPeriodicidadeMulta() {
        return this.periodicidadeMulta;
    }

    public BigDecimal getTaxa() {
        return this.taxa;
    }

    public BigDecimal getTetoTaxaDiaria() {
        return this.tetoTaxaDiaria;
    }

    public Boolean getPermiteReducao() {
        return this.permiteReducao;
    }

    public static TaxaMultaPrevidenciariaOptimizerListSearch obterListaOtimizada(Date dataInicio, Date dataFim) {
        TaxaMultaPrevidenciariaOptimizerListSearch taxaMultaPrevidenciariaOptimizerListSearch = new TaxaMultaPrevidenciariaOptimizerListSearch(dataFim);
        taxaMultaPrevidenciariaOptimizerListSearch.init((Collection<TaxaMultaPrevidenciaria>)TaxaMultaPrevidenciaria.getRepositorio(RepositorioDeTaxaMultaPrevidenciaria.class).obterTodos(dataInicio, dataFim));
        return taxaMultaPrevidenciariaOptimizerListSearch;
    }

    public static TaxaMultaPrevidenciaria obterTaxaNa(Date data) {
        List<TaxaMultaPrevidenciaria> lista = TaxaMultaPrevidenciaria.getRepositorio(RepositorioDeTaxaMultaPrevidenciaria.class).obterTodos(data, data);
        if (lista == null || lista.isEmpty()) {
            return null;
        }
        return lista.get(0);
    }

    public BigDecimal resolverTaxa(InssSobreSalarios inssSobreSalarios, Competencia competencia, boolean ehDecimoTerceiro, Date dataLiquidacao) {
        return this.resolverTaxa(inssSobreSalarios, competencia.getHelperDate(), ehDecimoTerceiro, dataLiquidacao, false);
    }

    public BigDecimal resolverTaxa(InssSobreSalarios inssSobreSalarios, HelperDate competenciaHelperDate, boolean ehDecimoTerceiro, Date dataLiquidacao, boolean semProjecaoDeData) {
        if (inssSobreSalarios.getInss().getCalculo().isCalculoExterno().booleanValue()) {
            return this.resolverTaxaCalculoExterno(inssSobreSalarios, dataLiquidacao, ehDecimoTerceiro);
        }
        Competencia competencia = Competencia.getInstance(competenciaHelperDate.getDate());
        ParametrosDeAtualizacao parametros = inssSobreSalarios.getInss().getCalculo().getParametrosDeAtualizacao();
        TipoPagamentoDaMultaDoINSSEnum tipoPagamento = null;
        Boolean existeDataDeInicioDeAtualizacaoPrevidenciaria = null;
        Boolean existeDataDeInicioDeLei11941 = null;
        Date aPartirLei11941 = null;
        Boolean ehEmpregadoDomestico = null;
        Date dataInicioAtualizacaoPrevidenciaria = null;
        if (inssSobreSalarios instanceof InssSobreSalariosDevidos) {
            AtividadeEconomica atividadeSelecionada;
            tipoPagamento = parametros.getPagamentoDaMultaDosSalariosDevidosDoINSS();
            existeDataDeInicioDeAtualizacaoPrevidenciaria = Utils.naoNulo(parametros.getAplicarAteDosSalariosDevidosDoINSS());
            existeDataDeInicioDeLei11941 = parametros.getLei11941() != false && Utils.naoNulo(parametros.getApartirDeLei11941());
            aPartirLei11941 = HelperDate.getCurrentCompetence(parametros.getApartirDeLei11941()).getDate();
            dataInicioAtualizacaoPrevidenciaria = parametros.getAplicarAteDosSalariosDevidosDoINSS();
            ehEmpregadoDomestico = ((InssSobreSalariosDevidos)inssSobreSalarios).getApurarInssSegurado().booleanValue() ? (!TipoDeAliquotaDoSeguradoEnum.EMPREGADO_DOMESTICO.equals((Object)inssSobreSalarios.getInss().getTipoAliquotaSegurado()) ? Boolean.valueOf(false) : Boolean.valueOf(true)) : (!TipoDeAliquotaDoEmpregadorEnum.POR_ATIVIDADE_ECONOMICA.equals((Object)inssSobreSalarios.getInss().getTipoAliquotaEmpregador()) ? Boolean.valueOf(false) : (SERVICOS_DOMESTICOS.equals((atividadeSelecionada = inssSobreSalarios.getInss().getAtividadeEconomica()).getDescricao().toUpperCase()) ? Boolean.valueOf(true) : Boolean.valueOf(false)));
        } else {
            tipoPagamento = parametros.getPagamentoDaMultaDosSalariosPagosDoINSS();
            existeDataDeInicioDeAtualizacaoPrevidenciaria = Utils.naoNulo(parametros.getAplicarAteDosSalariosPagosDoINSS());
            existeDataDeInicioDeLei11941 = parametros.getLei11941Pago() != false && Utils.naoNulo(parametros.getApartirDeLei11941Pago());
            aPartirLei11941 = HelperDate.getCurrentCompetence(parametros.getApartirDeLei11941Pago()).getDate();
            dataInicioAtualizacaoPrevidenciaria = parametros.getAplicarAteDosSalariosPagosDoINSS();
            ehEmpregadoDomestico = !TipoDeAliquotaDoSeguradoEnum.EMPREGADO_DOMESTICO.equals((Object)inssSobreSalarios.getInss().getTipoAliquotaSegurado()) ? Boolean.valueOf(false) : Boolean.valueOf(true);
        }
        switch (this.getPeriodicidadeMulta()) {
            case TAXA_UNICA: {
                if (HelperDate.dateEquals(competencia.getData(), dataLiquidacao)) {
                    return BigDecimal.ZERO;
                }
                if (TipoPagamentoDaMultaDoINSSEnum.REDUZIDO.equals((Object)tipoPagamento) && this.getPermiteReducao().booleanValue()) {
                    return this.getTaxa().divide(new BigDecimal(2), Utils.CONTEXTO_MATEMATICO);
                }
                return this.getTaxa();
            }
            case DIARIA: {
                BigDecimal taxa;
                HelperDate dataInicial = null;
                if (semProjecaoDeData) {
                    dataInicial = competenciaHelperDate.addDay(-1);
                } else if (!(!Boolean.TRUE.equals(existeDataDeInicioDeAtualizacaoPrevidenciaria) || existeDataDeInicioDeLei11941.booleanValue() && HelperDate.dateAfterOrEquals(competencia.getData(), aPartirLei11941))) {
                    Competencia competenciaInicialParaContagem = new Competencia(competencia.getMes(), competencia.getAno());
                    HelperDate competenciaAplicarAte = HelperDate.getCurrentCompetence(dataInicioAtualizacaoPrevidenciaria);
                    if (Utils.naoNulo(dataInicioAtualizacaoPrevidenciaria) && HelperDate.dateAfter(competenciaAplicarAte.getDate(), competencia.getData())) {
                        competenciaInicialParaContagem.setAno(competenciaAplicarAte.getYear());
                        competenciaInicialParaContagem.setMes(competenciaAplicarAte.getMonth());
                    }
                    dataInicial = HelperDate.getInstance(competenciaInicialParaContagem.getAno(), competenciaInicialParaContagem.getMes(), 2).addMonth(1);
                    while (!dataInicial.isWorkDayWithoutSaturdays()) {
                        dataInicial.addDay(1);
                    }
                } else if (ehEmpregadoDomestico.booleanValue() && !ehDecimoTerceiro) {
                    Competencia competenciaJunho2015 = new Competencia(5, 2015);
                    if (competencia.isAnteriorA(competenciaJunho2015.getData())) {
                        dataInicial = HelperDate.getInstance(competencia.getAno(), competencia.getMes(), 15).addMonth(1);
                        while (!dataInicial.isWorkDayWithoutSaturdays()) {
                            dataInicial.addDay(1);
                        }
                    } else {
                        dataInicial = HelperDate.getInstance(competencia.getAno(), competencia.getMes(), 7).addMonth(1);
                        while (!dataInicial.isWorkDayWithoutSaturdays()) {
                            dataInicial.addDay(-1);
                        }
                    }
                } else {
                    dataInicial = HelperDate.getInstance(competencia.getAno(), competencia.getMes(), 20).addMonth(1);
                    while (!dataInicial.isWorkDayWithoutSaturdays()) {
                        dataInicial.addDay(-1);
                    }
                }
                long totalDeDias = HelperDate.countDays(dataInicial.getDate(), dataLiquidacao);
                if (totalDeDias < 0L) {
                    totalDeDias = 0L;
                }
                if ((taxa = this.getTaxa().multiply(new BigDecimal(totalDeDias), Utils.CONTEXTO_MATEMATICO)).compareTo(this.getTetoTaxaDiaria()) > 0) {
                    taxa = this.getTetoTaxaDiaria();
                }
                return taxa;
            }
        }
        return null;
    }

    private BigDecimal resolverTaxaCalculoExterno(InssSobreSalarios inssSobreSalarios, Date dataLiquidacao, boolean ehDecimoTerceiro) {
        BigDecimal taxa;
        Calculo calculo = inssSobreSalarios.getInss().getCalculo();
        ParcelasAtualizaveisOutrosDebitosReclamado parcelasExternoOutrosDebitosReclamado = ParcelasAtualizaveisOutrosDebitosReclamado.obterDoCalculo(calculo);
        Date dataInicial = null;
        if (inssSobreSalarios instanceof InssSobreSalariosDevidos && calculo.getParametrosDeAtualizacao().getLei11941().booleanValue() && ehDecimoTerceiro) {
            dataInicial = parcelasExternoOutrosDebitosReclamado.getDataInicialAposFev2009MultaContribSocialDevidos();
        } else if (inssSobreSalarios instanceof InssSobreSalariosDevidos && calculo.getParametrosDeAtualizacao().getLei11941().booleanValue() && !ehDecimoTerceiro) {
            dataInicial = parcelasExternoOutrosDebitosReclamado.getDataInicialAteFev2009MultaContribSocialDevidos();
        } else if (inssSobreSalarios instanceof InssSobreSalariosDevidos && !calculo.getParametrosDeAtualizacao().getLei11941().booleanValue()) {
            dataInicial = parcelasExternoOutrosDebitosReclamado.getDataInicialBaseMultaContribSocialDevidos();
        } else if (inssSobreSalarios instanceof InssSobreSalariosPagos && calculo.getParametrosDeAtualizacao().getLei11941Pago().booleanValue() && ehDecimoTerceiro) {
            dataInicial = parcelasExternoOutrosDebitosReclamado.getDataInicialAposFev2009MultaContribSocialPagos();
        } else if (inssSobreSalarios instanceof InssSobreSalariosPagos && calculo.getParametrosDeAtualizacao().getLei11941Pago().booleanValue() && !ehDecimoTerceiro) {
            dataInicial = parcelasExternoOutrosDebitosReclamado.getDataInicialAteFev2009MultaContribSocialPagos();
        } else if (inssSobreSalarios instanceof InssSobreSalariosPagos && !calculo.getParametrosDeAtualizacao().getLei11941Pago().booleanValue()) {
            dataInicial = parcelasExternoOutrosDebitosReclamado.getDataInicialBaseMultaContribSocialPagos();
        }
        if (Utils.nulo(dataInicial)) {
            return BigDecimal.ZERO;
        }
        long totalDeDias = HelperDate.countDays(dataInicial, dataLiquidacao);
        if (totalDeDias < 0L) {
            totalDeDias = 0L;
        }
        if ((taxa = TAXA_MULTA_CALCULO_EXTERNO.multiply(new BigDecimal(totalDeDias), Utils.CONTEXTO_MATEMATICO)).compareTo(TETO_MULTA_CALCULO_EXTERNO) > 0) {
            taxa = TETO_MULTA_CALCULO_EXTERNO;
        }
        return taxa;
    }

    public BigDecimal resolverTaxaIrpf(Competencia competencia, Date dataLiquidacao) {
        long totalDeDias;
        HelperDate dataInicial = null;
        if (HelperDate.getInstance(dataLiquidacao).lessThanOrEqualsTo(TRINTA_SETEMBRO_2008)) {
            dataInicial = HelperDate.getInstance(competencia.getAno(), competencia.getMes(), 10).addMonth(1);
            while (!dataInicial.isWorkDayWithoutSaturdays()) {
                dataInicial.addDay(1);
            }
        } else {
            dataInicial = HelperDate.getInstance(competencia.getAno(), competencia.getMes(), 20).addMonth(1);
            while (!dataInicial.isWorkDayWithoutSaturdays()) {
                dataInicial.addDay(-1);
            }
        }
        if ((totalDeDias = HelperDate.countDays(dataInicial.getDate(), dataLiquidacao)) < 0L) {
            totalDeDias = 0L;
        }
        BigDecimal taxa = this.getTaxa().multiply(new BigDecimal(totalDeDias), Utils.CONTEXTO_MATEMATICO);
        if (Utils.naoNulo(this.getTetoTaxaDiaria()) && taxa.compareTo(this.getTetoTaxaDiaria()) > 0) {
            taxa = this.getTetoTaxaDiaria();
        }
        return taxa;
    }
}

File: src/lib/pjecalc/core/dominio/inss/multa/taxa-multa-previdenciaria.ts
/**
 * PJe-Calc v2.15.1 — TaxaMultaPrevidenciaria
 * Porte 1:1 de: br.jus.trt8.pjecalc.negocio.dominio.inss.multa.TaxaMultaPrevidenciaria
 *
 * Ref Java: pjecalc-fonte/.../inss/multa/TaxaMultaPrevidenciaria.java (~288 LOC)
 *
 * Entidade que guarda uma taxa de multa previdenciária (percentual) aplicada
 * em um período (dataInicial → dataFinal). Usada por MaquinaDeCalculoDoInss
 * para compor a multa das ocorrências (Lei 8.212/1991 e Lei 11.941/2009).
 *
 * Tabela Supabase: `pjecalc_inss_multa` (2 linhas: Lei 8.212 e Lei 11.941).
 * O carregamento real fica em `obterListaOtimizada` (adapter Supabase).
 * Enquanto o adapter não é instalado via DI, usamos a fonte padrão que
 * retorna lista vazia — o chamador (MaquinaDeCalculoDoInss) já trata isso.
 */
import Decimal from 'decimal.js';
import { HelperDate } from '../../../base/comum/helper-date';
import { naoNulo, nulo } from '../../../base/comum/utils';
import { Competencia } from '../../../base/comum/competencia';
import {
  TipoDaMultaDoINSSEnum,
  TipoDeAliquotaDoEmpregadorEnum,
  TipoDeAliquotaDoSeguradoEnum,
  TipoPagamentoDaMultaDoINSSEnum,
  TipoPeriodicidadeMultaPrevidenciariaEnum,
} from '../../../constantes/enums';

/** Duck type mínimo para InssSobreSalarios — evita dependência circular. */
export interface InssSobreSalariosRef {
  getInss(): {
    getCalculo(): {
      getCalculoExterno(): boolean;
      getParametrosDeAtualizacao(): {
        getPagamentoDaMultaDosSalariosDevidosDoINSS(): TipoPagamentoDaMultaDoINSSEnum;
        getPagamentoDaMultaDosSalariosPagosDoINSS(): TipoPagamentoDaMultaDoINSSEnum;
        getAplicarAteDosSalariosDevidosDoINSS(): Date | null;
        getAplicarAteDosSalariosPagosDoINSS(): Date | null;
        getLei11941(): boolean;
        getLei11941Pago(): boolean;
        getApartirDeLei11941(): Date | null;
        getApartirDeLei11941Pago(): Date | null;
      };
    };
    getTipoAliquotaSegurado(): TipoDeAliquotaDoSeguradoEnum;
    getTipoAliquotaEmpregador(): TipoDeAliquotaDoEmpregadorEnum;
    getAtividadeEconomica(): { getDescricao(): string } | null;
  };
  isDevidos(): boolean;
  getApurarInssSegurado?(): boolean;
}

const SERVICOS_DOMESTICOS = 'SERVIÇOS DOMÉSTICOS';
const TAXA_MULTA_CALCULO_EXTERNO = new Decimal('0.33');
const TETO_MULTA_CALCULO_EXTERNO = new Decimal(20);
const TRINTA_SETEMBRO_2008 = new Date(2008, 8, 30);

/** Mon-Fri considerado dia útil (sem feriados nacionais). */
function isWorkDayWithoutSaturdays(data: Date): boolean {
  const dow = data.getDay();
  return dow >= 1 && dow <= 5;
}

/** Avança `hd` até cair em dia útil (direção +1 ou -1). */
function ajustarParaDiaUtil(hd: HelperDate, direcao: 1 | -1): HelperDate {
  let guard = 0;
  while (!isWorkDayWithoutSaturdays(hd.getDate()) && guard++ < 14) {
    hd.addDay(direcao);
  }
  return hd;
}

export class TaxaMultaPrevidenciaria {
  private id: number | null = null;
  private dataInicial: Date;
  private dataFinal: Date | null;
  private tipoMulta: TipoDaMultaDoINSSEnum = TipoDaMultaDoINSSEnum.URBANA;
  private periodicidadeMulta: TipoPeriodicidadeMultaPrevidenciariaEnum = TipoPeriodicidadeMultaPrevidenciariaEnum.TAXA_UNICA;
  private taxa: Decimal;
  private tetoTaxaDiaria: Decimal | null;
  private permiteReducao: boolean = false;

  constructor(
    dataInicial: Date,
    dataFinal: Date | null,
    tipoMulta: TipoDaMultaDoINSSEnum,
    periodicidadeMulta: TipoPeriodicidadeMultaPrevidenciariaEnum,
    taxa: Decimal,
    tetoTaxaDiaria: Decimal | null = null,
    permiteReducao: boolean = false,
  ) {
    this.dataInicial = dataInicial;
    this.dataFinal = dataFinal;
    this.tipoMulta = tipoMulta;
    this.periodicidadeMulta = periodicidadeMulta;
    this.taxa = taxa;
    this.tetoTaxaDiaria = tetoTaxaDiaria;
    this.permiteReducao = permiteReducao;
  }

  getId(): number | null { return this.id; }
  setId(v: number | null): void { this.id = v; }

  getDataInicial(): Date { return this.dataInicial; }
  getDataFinal(): Date | null { return this.dataFinal; }
  getTipoMulta(): TipoDaMultaDoINSSEnum { return this.tipoMulta; }
  getPeriodicidadeMulta(): TipoPeriodicidadeMultaPrevidenciariaEnum { return this.periodicidadeMulta; }
  getTaxa(): Decimal { return this.taxa; }
  getTetoTaxaDiaria(): Decimal | null { return this.tetoTaxaDiaria; }
  getPermiteReducao(): boolean { return this.permiteReducao; }

  /**
   * resolverTaxa (Java linha 142 / 146).
   *
   * Calcula a taxa aplicável à competência, considerando:
   * - cálculo externo (Lei 11.941 com contagem de dias)
   * - periodicidade TAXA_UNICA (mesma taxa) ou DIARIA (multiplicada por dias)
   * - redução (Lei 8.212: reduzida em 50%)
   */
  resolverTaxa(
    inssSobreSalarios: InssSobreSalariosRef,
    competenciaHelperDate: HelperDate,
    ehDecimoTerceiro: boolean,
    dataLiquidacao: Date,
    semProjecaoDeData: boolean = false,
  ): Decimal | null {
    const calculo = inssSobreSalarios.getInss().getCalculo();
    if (calculo.getCalculoExterno()) {
      // TODO(integracao-futura): resolverTaxaCalculoExterno quando
      // ParcelasAtualizaveisOutrosDebitosReclamado estiver integrado ao fluxo.
      return new Decimal(0);
    }

    const competencia = Competencia.getInstance(competenciaHelperDate.getDate());
    const parametros = calculo.getParametrosDeAtualizacao();

    const ehDevidos = inssSobreSalarios.isDevidos();
    const tipoPagamento = ehDevidos
      ? parametros.getPagamentoDaMultaDosSalariosDevidosDoINSS()
      : parametros.getPagamentoDaMultaDosSalariosPagosDoINSS();

    const dataInicioAtualizacaoPrevidenciaria = ehDevidos
      ? parametros.getAplicarAteDosSalariosDevidosDoINSS()
      : parametros.getAplicarAteDosSalariosPagosDoINSS();
    const existeDataDeInicioDeAtualizacaoPrevidenciaria = naoNulo(dataInicioAtualizacaoPrevidenciaria);

    const apartirDeLei11941 = ehDevidos
      ? parametros.getApartirDeLei11941()
      : parametros.getApartirDeLei11941Pago();
    const aplicarLei11941 = ehDevidos ? parametros.getLei11941() : parametros.getLei11941Pago();
    const existeDataDeInicioDeLei11941 = aplicarLei11941 && naoNulo(apartirDeLei11941);
    const aPartirLei11941 = apartirDeLei11941
      ? HelperDate.getCurrentCompetence(apartirDeLei11941).getDate()
      : null;

    const inss = inssSobreSalarios.getInss();
    let ehEmpregadoDomestico = false;
    if (ehDevidos) {
      const apurarSegurado = inssSobreSalarios.getApurarInssSegurado?.() ?? true;
      if (apurarSegurado) {
        ehEmpregadoDomestico =
          inss.getTipoAliquotaSegurado() === TipoDeAliquotaDoSeguradoEnum.EMPREGADO_DOMESTICO;
      } else if (inss.getTipoAliquotaEmpregador() === TipoDeAliquotaDoEmpregadorEnum.POR_ATIVIDADE_ECONOMICA) {
        const desc = inss.getAtividadeEconomica()?.getDescricao().toUpperCase() ?? '';
        ehEmpregadoDomestico = desc === SERVICOS_DOMESTICOS;
      }
    } else {
      ehEmpregadoDomestico =
        inss.getTipoAliquotaSegurado() === TipoDeAliquotaDoSeguradoEnum.EMPREGADO_DOMESTICO;
    }

    switch (this.periodicidadeMulta) {
      case TipoPeriodicidadeMultaPrevidenciariaEnum.TAXA_UNICA: {
        const compData = competencia.getData();
        if (compData && HelperDate.dateEquals(compData, dataLiquidacao)) return new Decimal(0);
        if (tipoPagamento === TipoPagamentoDaMultaDoINSSEnum.REDUZIDO && this.permiteReducao) {
          return this.taxa.div(2);
        }
        return this.taxa;
      }
      case TipoPeriodicidadeMultaPrevidenciariaEnum.DIARIA: {
        let dataInicial: HelperDate;
        if (semProjecaoDeData) {
          dataInicial = competenciaHelperDate.addDay(-1);
        } else if (
          existeDataDeInicioDeAtualizacaoPrevidenciaria &&
          !(existeDataDeInicioDeLei11941 && aPartirLei11941 && compData(competencia) && HelperDate.dateAfterOrEquals(compData(competencia)!, aPartirLei11941))
        ) {
          const competenciaInicialParaContagem = new Competencia({
            mes: competencia.getMes() ?? 1,
            ano: competencia.getAno() ?? 1970,
          });
          if (dataInicioAtualizacaoPrevidenciaria) {
            const competenciaAplicarAte = HelperDate.getCurrentCompetence(dataInicioAtualizacaoPrevidenciaria);
            const compAtualData = compData(competencia);
            if (compAtualData && HelperDate.dateAfter(competenciaAplicarAte.getDate(), compAtualData)) {
              competenciaInicialParaContagem.setAno(competenciaAplicarAte.getYear());
              // Java usa getMonth() retornando 0-indexed — ajustamos para 1-indexed do Competencia.
              competenciaInicialParaContagem.setMes(competenciaAplicarAte.getMonth() + 1);
            }
          }
          dataInicial = HelperDate.getInstance(
            competenciaInicialParaContagem.getAno() ?? 1970,
            (competenciaInicialParaContagem.getMes() ?? 1) - 1,
            2,
          ).addMonth(1);
          ajustarParaDiaUtil(dataInicial, 1);
        } else if (ehEmpregadoDomestico && !ehDecimoTerceiro) {
          const competenciaJunho2015 = new Date(2015, 5, 1);
          const compAtualData = compData(competencia);
          if (compAtualData && compAtualData < competenciaJunho2015) {
            dataInicial = HelperDate.getInstance(
              competencia.getAno() ?? 1970,
              (competencia.getMes() ?? 1) - 1,
              15,
            ).addMonth(1);
            ajustarParaDiaUtil(dataInicial, 1);
          } else {
            dataInicial = HelperDate.getInstance(
              competencia.getAno() ?? 1970,
              (competencia.getMes() ?? 1) - 1,
              7,
            ).addMonth(1);
            ajustarParaDiaUtil(dataInicial, -1);
          }
        } else {
          dataInicial = HelperDate.getInstance(
            competencia.getAno() ?? 1970,
            (competencia.getMes() ?? 1) - 1,
            20,
          ).addMonth(1);
          ajustarParaDiaUtil(dataInicial, -1);
        }

        let totalDeDias = HelperDate.countDays(dataInicial.getDate(), dataLiquidacao);
        if (totalDeDias < 0) totalDeDias = 0;
        let taxa = this.taxa.mul(totalDeDias);
        if (this.tetoTaxaDiaria && taxa.greaterThan(this.tetoTaxaDiaria)) taxa = this.tetoTaxaDiaria;
        return taxa;
      }
    }
    return null;
  }

  /** resolverTaxaIrpf (Java linha 263) — multa mensal de IRPF. */
  resolverTaxaIrpf(competencia: Competencia, dataLiquidacao: Date): Decimal | null {
    const hdLiquidacao = HelperDate.getInstance(dataLiquidacao);
    const hdTrinta = HelperDate.getInstance(TRINTA_SETEMBRO_2008);
    let dataInicial: HelperDate;
    if (hdLiquidacao.getDate() <= hdTrinta.getDate()) {
      dataInicial = HelperDate.getInstance(
        competencia.getAno() ?? 1970,
        (competencia.getMes() ?? 1) - 1,
        10,
      ).addMonth(1);
      ajustarParaDiaUtil(dataInicial, 1);
    } else {
      dataInicial = HelperDate.getInstance(
        competencia.getAno() ?? 1970,
        (competencia.getMes() ?? 1) - 1,
        20,
      ).addMonth(1);
      ajustarParaDiaUtil(dataInicial, -1);
    }
    let totalDeDias = HelperDate.countDays(dataInicial.getDate(), dataLiquidacao);
    if (totalDeDias < 0) totalDeDias = 0;
    let taxa = this.taxa.mul(totalDeDias);
    if (naoNulo(this.tetoTaxaDiaria) && taxa.greaterThan(this.tetoTaxaDiaria)) taxa = this.tetoTaxaDiaria;
    if (nulo(taxa)) return null;
    return taxa;
  }
}

/** Helper tipado para pegar data da competência sem nullability boilerplate. */
function compData(c: Competencia): Date | null {
  return c.getData();
}
